### PR TITLE
Rewrite publish logic

### PR DESCRIPTION
**Pull request by confused-Techie**
After numerous issues and reports and pain points of this publish logic, we hit a point after enabling strict versioning on the backend to discover that the original logic completely fails now.

I'll try to update this with more details with more time, but for now, here's an attempt that should resolve our woes.

### CODE REVIEW

**Comment by DeeDeeG** [comment]
Posting this feedback from the Discord thread, in case it gets lost otherwise:

---

`ppm publish --tag` isn't working with this branch:
```
% ~/ppm/bin/ppm publish --tag v0.3.0
A version, tag, or new package name is required
```
Working on `master` branch:
```
% ~/ppm/bin/ppm publish --tag v0.3.0
Publishing deedeeg-my-package-10982367019823@v0.3.0 ✓
```

Sadly it's seeming like we might not have any unit tests for this?

**Comment by DeeDeeG** [comment]
Speaking of which, the Discord thread: https://discord.com/channels/992103415163396136/1241217301529886853

And the start of the recent round of discussion around this topic, preceding the formal Discord thread about it by a bit: https://discord.com/channels/992103415163396136/992109539346370661/1240836443299905606

File: src/publish.js [hunk 1]
```diff
@@ -440,56 +453,71 @@ have published it.\
         return error;
       }
 
-      if ((version?.length > 0) || (rename?.length > 0)) {
-        let originalName;
-        if (version?.length <= 0) { version = 'patch'; }
-        if (rename?.length > 0) { originalName = pack.name; }
+      if (version?.length === 0 && tag?.length === 0 && rename?.length === 0) {
```
Comment (DeeDeeG) [inline src/publish.js:456]: I think it would be a benefit to know this is somehow still ending up `undefined` (despite the normalization we're doing) sooner, due `undefined.length` giving an error `Uncaught TypeError: Cannot read properties of undefined (reading 'length')`.

```suggestion
      if (version.length === 0 && tag.length === 0 && rename.length === 0) {
```

Sort of theoretical, since it shouldn't be possible.

But also, I like to avoid more complicated syntax if it's not needed, for ease of reading.

Also: Given the head-scratcher this issue was to solve, correctness and surfacing error messages in as straight-forward a way as possible is kind of my top priority, rather than avoiding errors in this case. Unless we want to try/catch and print the error and exit gracefully, but... I don't think that helps anything anyway in this case.

**tl;dr: Hitting a simple error message would be better than not hitting it, in this case?**

Not a blocker, since it should all be hypothetical with all the normalizing done in this PR. But the goose-chase tracking down strange behavior makes me a tad paranoid, not gonna lie.

Comment (DeeDeeG) [reply src/publish.js:456]: Sorry for all the nitpicking, just inclined to make this code _very tidy_ if possible after what had to be thought through to get here.

Comment (DeeDeeG) [reply src/publish.js:456]: _Note: If moving away from optional chaining for the `varName.length` checks, there are three more in the diff for this PR, I just didn't individually comment them all._

Comment (DeeDeeG) [reply src/publish.js:456]: P.S. I suppose these instances of optional chaining mirror what was there before, but while we're addressing tiny subtleties of behavior that were unintuitive, I think it's worth deciding whether to move away from the optional chaining or not.

**Review by DeeDeeG — APPROVED** [review]
Giving the ol' "Approve with optional feedback".

Looks Good to Me, tested _thoroughly_ in the linked Discord thread. Worked for both of us on separate machines, should be in good shape now. 👍

File: src/publish.js [hunk 2]
```diff
@@ -359,7 +359,7 @@ have published it.\
           this.spawn(gitCommand, ['commit', '-m', message], (code, stderr, stdout) => {
             stderr ??= '';
             stdout ??= '';
-            if (code === 0) {
+            if (code !== 0) {
```
Comment (DeeDeeG) [inline src/publish.js:362]: Kind of odd this was wrong, I believe it was inadvertently changed as part of the _utterly massive_ promisification PRs, from looking at the blame history.

This change seems legit. All other calls to `logFailure()` are when `code !== 0`, as in error code rather than clean exit code. So yeah, seems legit IMO.

**Review by DeeDeeG — APPROVED** [review]
Approving, since this fixes the various intended use-cases, such as `ppm publish patch` for a first publish, while keeping `--tag` working, and fixing `--rename` which may have been broken for some time on `master` branch.

Might be a good idea either in this PR to do a follow-up to handle invalid combos of CLI args, such as doing `patch`, `--tag`, and `--rename` all at once? But I'm not sure that's actually invalid or if someone could intentionally do something useful with combining them? So, this is _optional_ to address, IMO. We can tell people to do just one at a time without having the code print an error message about it for now?

If you do limit to only one of them being set at a time, that could be reasonable, but honestly I think it's good enough for now without that as well.

File: src/publish.js (hunk 2 above)
Comment (confused-Techie) [reply src/publish.js:362]: Was surprised to see this one broken for so long tbh

**Comment by confused-Techie** [comment]
Thanks for the approval, gonna go ahead and merge this one for now, then we just gotta get it updated in pulsar sometime soon
